otherlibs/stdune/digest.ml

```diff
@@ -120,9 +122,32 @@ let path_with_stats ~allow_dirs path (stats : Stats_for_digest.t) :
       (file_with_executable_bit ~executable)
       path
     |> Path_digest_result.of_result
-  | S_DIR when allow_dirs ->
+  | S_DIR when allow_dirs -> (
     (* CR-someday amokhov: The current digesting scheme has collisions for files
        and directories. It's unclear if this is actually a problem. If it turns
        out to be a problem, we should include [st_kind] into both digests. *)
-    Ok (generic (stats.st_size, stats.st_perm, stats.st_mtime, stats.st_ctime))
+    match Path.readdir_unsorted path with
+    | Error e -> Path_digest_result.Unix_error e
+    | Ok listing -> (
+      match
+        List.rev_map listing ~f:(fun name ->
+            let path = Path.relative path name in
+            let stats =
+              match Path.stat path with
+              | Ok stat -> Stats_for_digest.of_unix_stats stat
+              | Error e -> raise_notrace (E e)
+            in
+            let digest = path_with_stats ~allow_dirs path stats in
```

Comment:
@snowleopard should we keep a set of visited nodes to guard against recursive symlinks?

Reply:
We return `Unexpected_kind` given a symlink (below), and I think we should do the same here, conservatively. You could just switch to `Path.lstat` above, and we won't need to think about it (for now).

Reply:
Looking at this again:

1. Broken symlinks aren't even a problem. `stat` will fail and the caller handles symlink loops already. I guess that means that directory targets (and dependencies) can't contain invalid symlinks - but that seems fine.

2. I don't see how the `Unexpected_kind` is possible. As far as I can see, we only call `path_with_stats stats` with `stats` obtained from `Unix.stat`. So I don't see how can the result ever be a symlink. To confirm, I made `S_LNK` pattern return `assert false` and confirmed that it's never triggered.

Reply:
My point is that `path_with_stats` (as written) doesn't handle symbolic links correctly if they appear in a directory. Digesting the destination isn't sufficient. Surely, two directories that have different symbolic links that just happen to point to destinations with the same digest shouldn't be considered identical.

The right thing to do is to digest the destination, as well as all the intermediate links you needed to follow on the way to it.

And, by the way, that only call site you found is probably just broken.

So, the simplest thing to do for now is to just refuse computing digests of directories containing symbolic links. We already had a return variant for that, so let's just use it. Later, if we do want to allow symbolic links inside digested directories, we can extend the functionality.

Reply:
> Surely, two directories that have different symbolic links that just happen to point to destinations with the same digest shouldn't be considered identical.

Good point. Will write a unit test for this.

> The right thing to do (I think) is to digest the destination, as well as all the intermediate links you needed to follow on the way to it.

Yes, that seems like the safe thing to do. However, if we are traversing the links manually, we indeed need a check for loops.